src/SlidingWindow.cpp

```diff
@@ -340,46 +488,128 @@ class SlidingWindowOnFunctionAndLoop : public IRMutator {
     }
 
 public:
-    SlidingWindowOnFunctionAndLoop(Function f, string v, Expr v_min)
-        : func(std::move(f)), loop_var(std::move(v)), loop_min(std::move(v_min)) {
+    SlidingWindowOnFunctionAndLoop(Function f, string v, Expr v_min, set<int> &slid_dimensions)
+        : func(std::move(f)), loop_var(std::move(v)), loop_min(std::move(v_min)), slid_dimensions(slid_dimensions) {
     }
+
+    Expr new_loop_min;
 };
 
-// Perform sliding window optimization for a particular function
-class SlidingWindowOnFunction : public IRMutator {
-    Function func;
+// In Stmt s, does the production of b depend on a?
+// We can't use produce/consume nodes to determine this, because they're "loose".
+// For example, we get this:
+//
+//  produce a {
+//   a(...) = ...
+//  }
+//  consume a {
+//   produce b {
+//    b(...) = ... // not depending on a
+//   }
+//   consume b {
+//    c(...) = a(...) + b(...)
+//   }
+//  }
+//
+// When we'd rather see this:
+//
+//  produce a {
+//   a(...) = ...
+//  }
+//  produce b {
+//   b(...) = ... // not depending on a
+//  }
+//  consume a {
+//   consume b {
+//    c(...) = a(...) + b(...)
+//   }
+//  }
+//
+// TODO: We might also need to figure out transitive dependencies...? If so, it
+// would be best to just fix the produce/consume relationships as above. We would
+// just be able to look for produce b inside produce a.
+class Dependencies : public IRVisitor {
+    using IRVisitor::visit;
 
-    using IRMutator::visit;
+    const string &producer;
+    bool in_producer = false;
 
-    Stmt visit(const For *op) override {
-        debug(3) << " Doing sliding window analysis over loop: " << op->name << "\n";
+    void visit(const ProducerConsumer *op) override {
+        ScopedValue<bool> old_finding_a(in_producer, in_producer || (op->is_producer && op->name == producer));
+        return IRVisitor::visit(op);
+    }
 
-        Stmt new_body = op->body;
+    void visit(const Call *op) override {
+        if (in_producer && op->call_type == Call::Halide) {
+            if (op->name != producer) {
+                dependencies.insert(op->name);
+            }
+        }
+        IRVisitor::visit(op);
+    }
+
+public:
+    set<string> dependencies;
 
-        new_body = mutate(new_body);
+    Dependencies(const string &producer)
+        : producer(producer) {
+    }
+};
 
-        if (op->for_type == ForType::Serial ||
-            op->for_type == ForType::Unrolled) {
-            new_body = SlidingWindowOnFunctionAndLoop(func, op->name, op->min).mutate(new_body);
+bool depends_on(const string &a, const string &b, const Stmt &s) {
+    if (a == b) {
+        return true;
+    }
+    Dependencies deps(b);
+    s.accept(&deps);
+    // Recursively search for dependencies. Repeatedly using this on the
+    // same set of Funcs is algorithmically slow, but even an absurd number
+    // of Funcs is still relatively small...
+    for (const string &i : deps.dependencies) {
```

Comment:
I think this is algorithmically slow enough to hit age of the universe for pathological pipelines. Can be fixed by adding a `map<pair<string, string>, bool>` arg and a cache check at the function entry point